File: 1039-find-the-town-judge/find-the-town-judge.cpp
class Solution {
public:
    int findJudge(int n, vector<vector<int>>& trust) {
        if(n == 1) return 1;
        map<int, vector<int>>m;
        map<int,vector<int>>rev;
        for(auto i : trust)
        {
            int first = i[0];
            int second = i[1];
            m[second].push_back(first);
            rev[first].push_back(second);
        }
        for(auto i : m)
        {
            int len = i.second.size();
            if(len == n-1 && rev.find(i.first) == rev.end())
            {
                return i.first;
            }
        }
        return -1;
    }
};

File: 1039-find-the-town-judge/find-the-town-judge.py
class Solution:
    def findJudge(self, n: int, trust: List[List[int]]) -> int:
        # This uses a concept called in degree 
        if n == 1:
            return 1
        d = {}
        d1 = {}
        for a in trust:
            ai = a[0]
            bi = a[1]
            if bi not in d:
                d[bi] = [ai]
            else:
                d[bi].append(ai)
            if ai not in d1:
                d1[ai] = [bi]
            else:
                d1[ai].append(bi)
        for person in d:
            if len(d[person]) == n-1 and person not in d1:
                return person
        return -1